File: Daily Challenge/January 2025/Course Schedule IV.cs
public class Solution {
  private Dictionary<(int, int), bool> _dp = new Dictionary<(int, int), bool>();
  private bool[] _isVisited;
  private int _prerequisite;

  private void DFS(int node, List<int>[] graph) {
    _isVisited[node] = true;
    _dp[(_prerequisite, node)] = true;

    foreach (var next in graph[node]) {
      if (!_isVisited[next]) {
        DFS(next, graph);
      }
    }
  }

  public IList<bool> CheckIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
    _dp.Clear();
    _isVisited = new bool[numCourses];
    var graph = new List<int>[numCourses];
    for (int i = 0; i < numCourses; i++) {
      graph[i] = new List<int>();
    }

    foreach (var prerequisite in prerequisites) {
      graph[prerequisite[0]].Add(prerequisite[1]);
    }

    for (int i = 0; i < numCourses; i++) {
      _prerequisite = i;
      DFS(i, graph);
      Array.Fill(_isVisited, false);
    }

    var result = new List<bool>();
    foreach (var query in queries) {
      result.Add(_dp.GetValueOrDefault((query[0], query[1]), false));
    }

    return result;
  }
}

File: Daily Challenge/January 2025/Course Schedule IV.cpp
class Solution {
  map<pair<int, int>, bool> dp;
  vector<bool> isVisited;
  int prerequisite = 0;

  void dfs(int node, vector<vector<int>>& graph) {
    isVisited[node] = true;
    dp[{prerequisite, node}] = true;
  
    for (int child : graph[node]) {
      if (!isVisited[child]) {
        dfs(child, graph);
      }
    }
  }

public:
  vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
    dp.clear();
    isVisited.assign(numCourses, false);
    vector<vector<int>> graph(numCourses);

    for (auto& p : prerequisites) {
      graph[p[0]].push_back(p[1]);
    }

    for (int i = 0; i < numCourses; i++) {
      prerequisite = i;
      dfs(i, graph);
      isVisited.assign(numCourses, false);
    }

    vector<bool> result;
    for (auto& q : queries) {
      result.push_back(dp[{q[0], q[1]}]);
    }

    return result;

  }
};